### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by leaanthony** [wailsapp/wails#1484] [Releases]
## v2.7
- Custom Protocol Handler
  https://github.com/wailsapp/wails/issues/1830

## v3
- Multiple Windows
  https://github.com/wailsapp/wails/issues/1480
- Tray Menus
  https://github.com/wailsapp/wails/issues/1521
- Support WebKitGTK 2.36+ for Linux
  https://github.com/wailsapp/wails/issues/1498
- Context Menus
  https://github.com/wailsapp/wails/issues/1835
- Start on login
  https://github.com/wailsapp/wails/issues/1789
- Transparent Backgrounds
  https://github.com/wailsapp/wails/issues/1296
- Native Drag and Drop
  https://github.com/wailsapp/wails/issues/1090
- Single Instance of Application
  https://github.com/wailsapp/wails/issues/1351
- Notification Support
  https://github.com/wailsapp/wails/issues/1788

## v3.1.0
- Proxy Support
  https://github.com/wailsapp/wails/issues/1469
- Self Updating
  https://github.com/wailsapp/wails/issues/1178

## v3.2.0
- More Menu Roles
  https://github.com/wailsapp/wails/issues/1130

## v4
- Support Android
  https://github.com/wailsapp/wails/issues/1481
- Support iOS
  https://github.com/wailsapp/wails/issues/1482

**Comment by daodao97** [comment]
This is encouraging, and a clear timeline would be even better. 👍🏻

**Reply by leaanthony** [reply]
A timeline is impossible. The only guarantee is that the more people who contribute the sooner it will happen.

**Reply by daodao97** [reply]
> A timeline is impossible. The only guarantee is that the more people who contribute the sooner it will happen.

I am very happy to contribute code, my technology stack is Golang/JS, but the native knowledge is a blind spot, I think many people are like me, I would like to know what can I do for Wails in this situation.

**Reply by leaanthony** [reply]
Thank you for your offer! There are many opportunities to contribute. For example, testing takes more time than developing. I'm going to formalise a workflow as part of the v2 release to better enable contributors like yourself. Thanks again 🙏

**Reply by steowens** [reply]
I would love to contribute but might need a little guidance to get off to a good start.  But if it's testing you need I am down with that. I am building an app with this thing now and I am sure to find many issues since I intend to put it through it's paces.

**Reply by leaanthony** [reply]
Appreciate the offer! Testing is 💯 a fantastic way to contribute to the project. A good way to get involved is looking at the PRs that get opened and testing them. Another way is looking at bug reports coming in and see if it's reproducible. If you fancy coding something, feel free to pick up a "good first issue" ticket 👍

**Reply by steowens** [reply]
I don't know about other people on this thread.  But if I were generous enough to devote my time to deliver something for free and someone asked me to provide a timeline on when I would deliver new features, I would say how much are you willing to pay for that timeline?  If your answer is nothing I would reply well you get what you pay for then.

**Reply by steowens** [reply]
Thanks leaanthony.  I was wondering if it might be possible to start by raising issues that new users (such as myself) are likely to encounter, such as an issue today.  It concerns taking a perfectly functioning version of the app I was so excited about and turning it in to something that does not compile simply by running go tidy.  Perhaps we could figure out how to get an app back in to a state where it works and I could add to the documentation recovery steps?  I am not sure I am ready to start testing PRs given my newness to the project.  For example I don't even know how to find the PRs.  When I click on Pull Requests in Github I get an empty search result.  OH, never mind I think I found them.

**Reply by leaanthony** [reply]
I guess Wails isn't currently in a state to just drop an app in and expect it to work. I mean we could make a window and that'd be about it. Have a think about what that might look like. I have some things on the back burner that might help with this

**Reply by steowens** [reply]
I suppose but I don't think it is too far off.  I have been reading through the Wails code and it seems like were almost there.  I have wails version 2..2.0.  At any rate I did a little side project just using webview to see how I could open up 2 windows. here:  https://github.com/steowens/goWebView.  At first when I did this I didn't pass the HINSTANCE of one into the create of another and essentially created 2 top level windows.  This was bad.  When I closed one of them the other would just hang and the entire process had to be killed.  BUT, if I only create ONE top level window then I can create the second window as a child and it just displays another window but now I can close the child without ruining the parent.

Now digging through your code was a bit of a head scratcher.  I got all the way down to the app package to the CreateApp function and became quite befuddled.  I cannot for the life of me see how that bit of code actually creates a browser container with menus.

At any rate, when you do wails Run 

    // Run creates an application based on the given config and executes it
    func Run(options *options.App) error {
    	mainApp := application.NewWithOptions(options)
    	return mainApp.Run()
    }

I see that you are throwing away the pointer to the new application but I think it would be a great idea to return that pointer to the caller.  Run could also include an optional pointer to a parent app so that when the child app starts it becomes a child window like the way I do in the raw webview things in my project link above.

**Reply by leaanthony** [reply]
The codebase is a little mixed up at the moment. There is a lot of tidying up to do and a need to move to a better API for multi window & multi tray. If you're keen on helping out, jump on discord and we can chat better there 👍 The discord link should be in slack if you can't find it.

**Reply by steowens** [reply]
Well I am starting to wrap my head around this where would I look to get on the discord channel?  

Never mind, I downloaded the app and then clicked on forgot my password and got stuck in an interminable loop of captcha exercises.  Not an app that the gods want me to use.  It's horrible experiences like the one Discord provides that I am trying to improve upon. 

Ok so they finally sent me a password reset link and I am on Discord.  But how do I find you?  I guess we need to friend up so here is my tag or whatever discord calls it: messageDelivered#2838

**Reply by leaanthony** [reply]
Try this https://discord.gg/WrtkKwWK

**Reply by steowens** [reply]
That worked thanks!

**Comment by aneeskA** [comment]
@leaanthony If I may, can https://github.com/wailsapp/wails/issues/1480 be included in 2.1?

**Reply by leaanthony** [reply]
There's a lot to talk about with multiple windows - it's a non trivial task.

**Reply by aneeskA** [reply]
> There's a lot to talk about with multiple windows - it's a non trivial task.

@leaanthony 
I understand. Let me explain why I am very interested in this particular feature. I work with Kafka streams day in day out. The tool I use - Conduktor - which started as an excellent tool (fast, efficient) two years is now a bloated mess - slow to load, slow to react etc. This was slowing me down at work. So I started looking into how I could write a performant desktop client for Kafka management. I decided to use librdkafka on Flutter. But I needed to learn a lot of new tech and figure out how to use C library efficiently in Flutter. I know I don't have that much time to dedicate right now.

That is when I stumbled across wails. The moment I saw I can write my business logic in Go, I couldn't contain my glee. I plan to spend more time building stuff using wails so that when the multi window feature rolls out, I will be ready.

Thanks a lot for your effort. This is going to save a lot of dev time.

**Reply by leaanthony** [reply]
I also work with Kafka so interested to see what you come up with. Is there a particular problem multiple windows solve that simple routing couldn't?

**Reply by aneeskA** [reply]
> I also work with Kafka so interested to see what you come up with. Is there a particular problem multiple windows solve that simple routing couldn't?

Yes. Here is what I usually do: There is a sender to a topic and there are 10 different topics that data gets produced into after various processing including error, stats etc. So I use one window to produce to the topic(sometimes in a loop), and then use separate windows to consume from the topics. This way, I can use the three finger swipe down - "App Expose" (in macOS trackpad) to see what is going on without leaving a single window. IMHO this makes it easy to grasp what happened at a glance instead of using the same window to consume to see what went down sequentially. 

Hope this makes sense to you.

**Reply by leaanthony** [reply]
What do you think a good API would be for that?

**Reply by steowens** [reply]
Not sure about the specifics for Kafka but in general, having a way for the main window and the child windows to send messages to each other would be a good general purpose API.  Each window could choose how to respond to those messages as it sees fit, and the Main Window would always be able to kill it's child windows if it sees fit.  Probably not too different than the API that Kafka itself implements, publish and subscribe.  Given that it's go, Channels seem the most appropriate, so if a window could be wrapped with a Go construct that is capable of sitting at the either end of channels that might be pretty darned cool.  If you wanted the child windows to be able to communicate with each other they could do it in pub-sub fashion brokered by the parent window.

So basically what I would imagine is that there is a general object lets call it Window.  It is responsible for setting up a single window which is basically what the App does today.  Each window would be capable of implementing a message bus that it and it's children can connect to.  This would enable maximum creativity.  The app developer can design the message schemes that they find most suitable for their needs, and once you have achieved this abstraction then windows could launch to arbitrary depths. Entire trees of windows could interoperate in arbitrarily complex interactions. 

Under this paradigm launching an app would be simply launching a window from the main function.  Now digging in to the app code a bit I can see that it already takes a context object and context seems to implement the kind of functionality that one would want to build a message bus on.  So I guess the only question is can there exist a wrapper stuct + functions that does what the typical wails main function does today so you can achieve recursion?

It looks like almost all of the pieces are nearly in place in the App construct.  Couple things missing on options are start page, as an option, and I have no idea how an App actually launches the underlying window.  It's beyond me how this code ends up producing the browser window we ultimately see when the app starts:

    `// CreateApp creates the app!
    func CreateApp(_ *options.App) (*App, error) {
    	result := w32.MessageBox(0,
    		`Wails applications will not build without the correct build tags. 
    Please use "wails build" or press "OK" to open the documentation on how to use "go build"`,
	    	"Error",
    		w32.MB_ICONERROR|w32.MB_OKCANCEL)
    	if result == 1 {
    		exec.Command("rundll32", "url.dll,FileProtocolHandler", "https://wails.io/docs/guides/manual-builds").Start()
    	}
    	return nil, nil
    }`

**Comment by dezashibi** [comment]
This is so charming 💯💯💯

**Comment by arunksoman** [comment]
I believe there was a self-update feature was on roadmap. Does that one removed? I didn't saw that feature on 2.3 changelogs.

**Reply by stffabi** [reply]
It has been rescheduled for v3

**Comment by ybh20** [comment]
有没有更详细的教程，官网案例太少了，好多东西都不会用

**Reply by alanchenchen** [reply]
这个和tauri感觉定位不太一样，tauri封装了很多os api，wails需要你自己用go实现func然后bind给js。所以得需要自己写go代码，web只做交互和展示

**Reply by balck-paint** [reply]
功能比tauri弱不少，如果熟悉go可以考虑，但像这种框架后台语言完全可以使用自己熟悉的用，这个就当个浏览器壳子用了，这样跟这个交互也不是很多

**Comment by Mrkouhadi** [comment]
Going through this roadmap gives more motivation to learn Wails and Go if you haven't started yet :D  .
Thank you guys and good luck !

**Reply by BenoitBotton** [reply]
I have been using wails for almost a year now and love it. This actually makes me want to contribute to my first open source project.
Thank you @leaanthony and all contributors for creating this wonderful framework

**Reply by leaanthony** [reply]
So kind @BenoitBotton! You are always welcome to contribute and we will help you in any way we can.

**Comment by pavelbinar** [comment]
Hi @leaanthony , it seems that version 3 is mostly ready, which is great 🎉 . Is there any planned release date or ETA for it?

Perhaps some features could be shifted to version 3.1.0 to enable quicker shipping of the completed code?

**Reply by neneodonkor** [reply]
How come no one answered? I am also interested in knowing the ETA.

**Reply by pavelbinar** [reply]
First, let me start with my apology to @leaanthony . Questions like mine:

> Is there any planned release date or ETA?

are quite annoying. They don’t add any value to the discussion and come off as passive-aggressive. If someone has an annoying or stupid question, I believe it’s good practice to ignore them.

Please learn from my mistakes.

Regarding the Wails v3 release, the answer is simple: it will be done when it is ready. I believe there is no deadline, no budget, and no paying customer who can complain. Therefore, it will be done when it is done. To learn more about the progress and all the exciting new features, I recommend checking the Wails Discord channel `wails3` at https://discord.com/channels/1042734330029547630/1065600636294549514.

**Reply by leaanthony** [reply]
That's fine @pavelbinar - if we knew, we'd say! There's only a couple of issues I want to iron out then we'll get to beta. Just no idea how long that might take 😅

**Comment by zhuweiyou** [comment]
Will v3 support load remote urls?

**Reply by bytehub** [reply]
From the example code of v3, it seems to be supported, but it's unknown whether the page loaded from a remote URL injects Wails runtime JS.

**Comment by Nanyte25** [comment]
I am trying out Wails and I wish we built more tools in in my company that are crossplatform like this as our main stack is go and kubernetes.

**Comment by ho0xi** [comment]
冒昧的问一下，V3预计什么时候发布稳定版

**Reply by itshizhan** [reply]
被移动端拖累了，起码要个三五年

**Comment by klgd** [comment]
will v3 support custom protocol scheme?

**Comment by LauGau** [comment]
Can't wait for mobile support! 🚀

**Comment by xxlyuan** [comment]
When is file dragging to other applications supported

**Reply by xxlyuan** [reply]
https://electron.nodejs.cn/docs/latest/tutorial/native-file-drag-drop 
demo

**Comment by seniloper** [comment]
what's up ... android ?

**Comment by Creepios** [comment]
@leaanthony is this roadmap being updated? I guess v3.0 would be nearly done with the current state of the roadmap :D

**Reply by johnmaguire** [reply]
I hesitate to even comment, as I understand the pressures on an OSS maintainer and I very much appreciate all the work on Wails. That said, using "alpha" software for years is beginning to make us a little nervous, especially as the only releases are nightlies (so no stability guarantees, and we have to watch out for breaking changes.)

That said, lots of people are using v3 alpha in production, so maybe it's time to remove the alpha tag?

**Reply by leaanthony** [reply]
No Yeah, I get that. It's been mostly stable for a while, which is a good indicator. But removing that flag has some other serious implications and I need to make sure it can be supported in the manner it should be

**Comment by laoshaw** [comment]
With claude code, I hope this project can proceed even faster, to get android and ios support sooner :)
